### PR TITLE
STORM-2844: KafkaSpout Throws IllegalStateException After Committing …

### PR DESCRIPTION
**Pull request by hmcl**
…to Kafka When First Poll Strategy Set to EARLIEST

     - Check if commits to Kafka were committed by this topology to correctly enforce FirstPollOffsetStrategy
     - CommitMetadata includes topology id, and task id and threadName

### CODE REVIEW

**Comment by hmcl** [comment]
Master PR

**Comment by hmcl** [comment]
@srdo this is the master version.

**Comment by hmcl** [comment]
@srdo these are all the changes in 1.x-branch already squashed. It is ready to merge.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/OffsetAndMetadataStub.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class OffsetAndMetadataStub {
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/OffsetAndMetadataStub.java]: This class was removed from the 1.x version I think, so it shouldn't be here either.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutSingleTopicTest.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+import static org.mockito.Matchers.anyList;
+import static org.mockito.Matchers.anyString;
+
+import java.util.regex.Pattern;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.hamcrest.Matchers;
+
+public class KafkaSpoutSingleTopicTest extends KafkaSpoutAbstractTest {
+    @Captor
+    private ArgumentCaptor<Map<TopicPartition, OffsetAndMetadata>> commitCapture;
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutSingleTopicTest.java]: This hides an identical field in the superclass.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutSingleTopicTest.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+        //Advance the time and replay the failed tuple. 
+        reset(collector);
+        spout.nextTuple();
+        ArgumentCaptor<KafkaSpoutMessageId> failedIdReplayCaptor = ArgumentCaptor.forClass(KafkaSpoutMessageId.class);
+        verify(collector).emit(anyString(), anyList(), failedIdReplayCaptor.capture());
+
+        assertThat("Expected replay of failed tuple", failedIdReplayCaptor.getValue(), is(failedTuple));
+
+            /* Ack the tuple, and commit.
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutSingleTopicTest.java]: Nit: Indentation on this is wrong.

**Comment by hmcl** [comment]
@srdo Thanks for the review. I have addressed the last few things and squashed the commits right away. It is ready to merge.

**Pull request merged**